service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -0,0 +1,698 @@
+        }
+
+        serviceNeedDiffByChild.value.additions.forEach { addedServiceNeedId ->
+            processedServiceNeedIds.add(addedServiceNeedId)
+            handleNewEvakaServiceNeed(db, client, addedServiceNeedId)
+        }
+    }
+
+    // Handle fee data only -changes (no changes in service need). If there was any service need change, all related
+    // fee data has already been sent to / deleted from varda, so we filter those service needs out here
+    feeAndVoucherDiffsByServiceNeed.filterNot { processedServiceNeedIds.contains(it.key) }.entries.forEach {
+        val serviceNeedId = it.key
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) }
+        if (vardaServiceNeed != null) {
+            handleUpdatedEvakaServiceNeed(db, client, serviceNeedId)
+        } else {
+            handleNewEvakaServiceNeed(db, client, serviceNeedId)
+        }
+    }
+}
+
+fun handleDeletedEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) }
+        if (vardaServiceNeed != null) {
+            val errors = deleteServiceNeedDataFromVarda(client, vardaServiceNeed)
+            if (errors.isNotEmpty()) {
+                logger.error(errors.joinToString(","))
+                db.transaction { it.markServiceNeedUpdateFailed(serviceNeedId, errors) }
+            } else {
+                db.transaction { it.deleteVardaServiceNeedByEvakaServiceNeed(serviceNeedId) }
+            }
+        }
+    } catch (e: Exception) {
+        logger.error("VardaUpdate: manual check needed: something went from while trying to delete varda service need $serviceNeedId data: ${e.localizedMessage}")
+    }
+}
+
+fun handleUpdatedEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) } ?: throw Exception("VardaUpdate: cannot update service need $serviceNeedId: varda service need missing")
+        val deleteErrors = deleteServiceNeedDataFromVarda(client, vardaServiceNeed)
+        if (deleteErrors.isNotEmpty()) {
+            logger.error(deleteErrors.joinToString(","))
+            db.transaction {
+                it.markServiceNeedUpdateFailed(serviceNeedId, deleteErrors)
+            }
+        } else {
+            val evakaServiceNeed = db.read { it.getEvakaServiceNeedInfoForVarda(serviceNeedId) }
+            val addErrors = addServiceNeedDataToVarda(db, client, evakaServiceNeed, vardaServiceNeed)
+            if (addErrors.isNotEmpty()) {
+                logger.error(addErrors.joinToString(","))
+                db.transaction {
+                    it.markServiceNeedUpdateFailed(serviceNeedId, addErrors)
+                }
+            }
+        }
+        db.transaction { it.upsertVardaServiceNeed(vardaServiceNeed) }
```

Comment:
Just to make sure... Is it intentional that this is run even if delete or add fails?